core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -154,6 +161,147 @@ class ControllerApis(val requestChannel: RequestChannel,
+      if (duplicatedTopicNames.contains(name) || !topicNamesToResolve.add(name)) {
+        appendResponse(name, ZERO_UUID, new ApiError(INVALID_REQUEST, "Duplicate topic name."))
+        topicNamesToResolve.remove(name)
+        duplicatedTopicNames.add(name)
+      }
+    }
+
+    def maybeAppendToIdsToResolve(id: Uuid): Unit = {
+      if (duplicatedTopicIds.contains(id) || !topicIdsToResolve.add(id)) {
+        appendResponse(null, id, new ApiError(INVALID_REQUEST, "Duplicate topic ID."))
+        topicIdsToResolve.remove(id)
+        duplicatedTopicIds.add(id)
+      }
+    }
+
+    request.topicNames().iterator().asScala.foreach(maybeAppendToTopicNamesToResolve)
+
+    request.topics().iterator().asScala.foreach {
+      topic => if (topic.name() == null) {
+        if (topic.topicId.equals(ZERO_UUID)) {
+          appendResponse(null, ZERO_UUID, new ApiError(INVALID_REQUEST,
+            "Neither topic name nor id were specified."))
+        } else {
+          maybeAppendToIdsToResolve(topic.topicId())
+        }
+      } else {
+        if (topic.topicId().equals(ZERO_UUID)) {
+          maybeAppendToTopicNamesToResolve(topic.name())
+        } else {
+          appendResponse(topic.name(), topic.topicId(), new ApiError(INVALID_REQUEST,
+            "You may not specify both topic name and topic id."))
+        }
+      }
+    }
+
+    val idToName = new util.HashMap[Uuid, String]
+
+    def maybeAppendToIdToName(id: Uuid, name: String): Unit = {
+      if (duplicatedTopicIds.contains(id) || idToName.put(id, name) != null) {
+          appendResponse(name, id, new ApiError(INVALID_REQUEST,
+              "The same topic was specified by name and by id."))
+          idToName.remove(id)
+          duplicatedTopicIds.add(id)
+      }
+    }
+    controller.findTopicIds(topicNamesToResolve).get().asScala.foreach {
+      case (name, idOrError) => if (idOrError.isError) {
+        appendResponse(name, ZERO_UUID, idOrError.error())
+      } else {
+        maybeAppendToIdToName(idOrError.result(), name)
+      }
+    }
+    controller.findTopicNames(topicIdsToResolve).get().asScala.foreach {
+      case (id, nameOrError) => if (nameOrError.isError) {
+        appendResponse(null, id, nameOrError.error())
+      } else {
+        maybeAppendToIdToName(id, nameOrError.result())
+      }
+    }
+
+    if (!hasClusterAuth) {
+      val authorizedDescribeTopics = getDescribableTopics(idToName.values().asScala)
+      val authorizedDeleteTopics = getDeletableTopics(idToName.values().asScala)
+      val iterator = idToName.entrySet().iterator()
+      while (iterator.hasNext) {
+        val entry = iterator.next()
+        val topicName = entry.getValue
+        if (!authorizedDeleteTopics.contains(topicName)) {
+          val topicId = entry.getKey
+          if (!authorizedDescribeTopics.contains(topicName)) {
+            // If the user is not authorized to describe the topic, we pretend that it
```

Comment:
The error should be `TOPIC_AUTHORIZATION_FAILED` if the client does not have describe permission regardless of existence.

Reply:
I think there are 5 cases:
1. name provided, topic exists, can't delete, maybe can describe => TOPIC_AUTHORIZATION_FAILED
2. name provided, topic doesn't exist, can't delete, can describe => UNKNOWN_TOPIC_OR_PARTITION
3. name provided, topic doesn't exist, can't delete, can't describe => TOPIC_AUTHORIZATION_FAILED
4. id provided, topic exists, can't delete, maybe can describe => TOPIC_AUTHORIZATION_FAILED
5. id provided, topic doesn't exists, can't delete, maybe can describe => UNKNOWN_TOPIC_ID

Reply:
For case 4, are we exposing the topic exists by returning a different error in the case where we can't describe?

Reply:
You're right, the earlier list was not quite right. I've revised this a bit. It's now:

```
no name or id => INVALID_REQUEST

duplicate name or id => INVALID_REQUEST

can't resolve topic id => UNKNOWN_TOPIC_ID

can't locate topic name => UNKNOWN_TOPIC_OR_PARTITION

no delete permission, no describe permission => UNKNOWN_TOPIC_ID (if topic id was provided) or UNKNOWN_TOPIC_OR_PARTITION (if topic name was provided)

no delete permission, describe permission => TOPIC_AUTHORIZATION_FAILED with both name and id filled out correctly
```

The new code should implement this correctly...

Reply:
Oh sorry this just got changed due to https://issues.apache.org/jira/browse/KAFKA-12394
So the case of no delete permission, no describe permission, topic ID provided is now TOPIC_AUTHORIZATION_FAILED. This may have been what you had initially.